### DISCUSSION TITLE
Content Assist Issues

### DISCUSSION THREAD

**Discussion by edshaw42** [codefori/vscode-ibmi#1252] [Q&A] [answered]
I am not seeing content assist anywhere. I have watched the tutorial and read the docs and it's not in my settings page. Maybe I am missing a piece somewhere?

**Comment by worksofliam** [comment]
@edshaw42 

You've not really provided any context here. Content assist for what language?

**Comment by edshaw42** [comment]
My apologies. 

I was watching your video on IBMi Notebooks. You referred to the content assist for CLP in that video and actually have a whole independent video over content assist. 

But, when I go to the settings page, I do not see content assist anywhere. Possibly has been moved through an upgrade over time, or I am missing a piece to the puzzle. I would love to have the content assist for RPG, CLP, and SQL if available. 

Attaching a screenshot of my settings page current view. 

![image](https://user-images.githubusercontent.com/85707702/233736513-c4a7e811-0466-4be3-9bfe-c8433f2f21b4.png)

**Reply by NorskieRider** [reply]
That's the settings menu I get.

Unlike the video, I don't have a "MEMBER BROWSER" section, but I do have an "OBJECT BROWSER" section. Maybe that's the difference?

**Comment by worksofliam** [comment]
You likely want these two 😄 

* RPGLE: https://marketplace.visualstudio.com/items?itemName=HalcyonTechLtd.vscode-rpgle
* CL: https://marketplace.visualstudio.com/items?itemName=IBM.vscode-clle

**Reply by b-smith-23** [reply]
I installed the CL extension, but I still don't see the content assist as in your video.

**Comment by SJLennon** [comment]
@edshaw42 
@b-smith-23 
I had to create a CLLE program this weekend.  I could not get content assist to work either.  But read on...

I  have the CLLE language tools from IBM loaded. There is unfortunately no documentation on how to use it.

 The video says to use Ctrl+Enter but that just gives me a new line. Ctrl+Space just seems to give a list of the variables in the program. 

So I uninstalled the CLLE language tools, exited VSCODE and installed the CLLE language tools. (This may not be necessary, in retrospect.) 

After a bit of experimentation, this is what I have figured out. Space is the key!

Start with a brand new CLLE program and enter "pgm".  After a short delay you see this:
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/c8985664-20d7-4f1f-92f5-c94eec2b00d4)

If you select the PGM entry you get this:
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/c0ac2e7b-d667-4c4d-9633-d9aa02576de5)

Now enter a command, say "sndmsg",  **followed by a space** (_the space is important!_) and you see this:
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/2610a341-45e0-4b3c-87a9-675ddbfab31d)

Pick a keyword, say "tousr" and get this:
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/ca65982b-58d0-4695-89cc-1550f64da6c5)

No enter a space and you get this:
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/3d940c17-c11d-4f20-b998-0bf001cbed93)

Pick a value, then enter a space and you can build up the command:
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/2f001deb-8bc9-4903-99dc-01514f47c93f)

**Reply by b-smith-23** [reply]
I tried your suggestion and I am still not seeing the parameters.  When I type PGM I do get the option to select PGM and it adds the ENDPGM, but when I try the SNDMSG example followed by a space it doesn't do anything.  I tried uninstalling and reinstalling CLLE language tools, but that didn't help.  Not sure what is wrong here.  What could be different?  Here is the extensions that I have installed.

![image](https://github.com/codefori/vscode-ibmi/assets/150696892/35ab228d-e852-4a41-8974-031276d989fa)

**Reply by worksofliam** [reply]
Are you connected to a system when prompting?

**Comment by b-smith-23** [comment]
I believe so unless I am missing a step.  I have the source open from the system.

Disclaimer

The information contained in this communication from the sender is confidential. It is intended solely for use by the recipient and others authorized to receive it. If you are not the recipient, you are hereby notified that any disclosure, copying, distribution or taking action in relation of the contents of this information is strictly prohibited and may be unlawful.

This email has been scanned for viruses and malware, and may have been automatically archived by Mimecast Ltd, an innovator in Software as a Service (SaaS) for business. Providing a safer and more useful place for your human generated data. Specializing in; Security, archiving and compliance. To find out more visit the Mimecast website.

**Reply by SJLennon** [reply]
@b-smith-23 Don't know what else to say. 
But as mentioned above, you do have to be connected to the IBM i for this to work, because as far as I understand, it uses a program named GENCMDXML which is installed in your temporary library. Do you have a stable connection to the IBM i? Using a VPN which might cause problems?
Wonder if @edshaw42 managed to get it to work.

**Comment by SJLennon** [comment]
@worksofliam 
Further update.  **I can no longer get prompting to work this morning (Dec 15).**  
Last used date of GENCMDXML is 12/13/23.
There were various changes between 12/13 and 12/15.
CL Extension I am using was updated to version 1.1.3 on 12/15. I went back to 1.1.2 and it still does not work.
I updated VSCODE to 1.85.1 on 12/13. So I went back to 1.84.2.  Still no success.
I went back to Code for IBM i version 2.5.0 and prompting now works. Last used date of GENCMDXML is now 12/15/23.
Reinstalled VSCODE 1.85.1.  Prompting still works.
Updated CL to 1.1.3. Prompting still works.
Updated Code for IBM i to 2.6.1 which came out just a few hour ago. Prompting no longer works. Tried 2.6.0--doesn't work.
**So it looks like the update of Code for IBM i to version 2.6.0 broke something.**

**Reply by worksofliam** [reply]
I will take a look today. Thanks

**Reply by worksofliam** [reply]
Just released 2.6.1 which may have actually solved this.

**Reply by worksofliam** [reply]
Fixed in vscode-clle 1.1.4.

**Comment by SJLennon** [comment]
@worksofliam 
Sorry, 2.6.1 doesn't fix it.  I had already tried that. 2.5.0 seems to be the last one that works.  Fortunately, I don't write a lot of CL.
Just a reminder, I'm on PUB400.

**Reply by worksofliam** [reply, marked as answer]
Looks like I found the issue. We removed the API that vscode-clle uses. Working on a fix today!

<img width="678" alt="image" src="https://github.com/codefori/vscode-ibmi/assets/3708366/12261d83-fbbc-4351-aefd-4cb2d2258115">

edit: vscode-clle 1.1.4 is out and the issue has been fixed.

**Reply by b-smith-23** [reply]
Yes, it's working now!  Thank you!

**Comment by b-smith-23** [comment]
Great!  

While you're at it, I discovered a minor bug in the CRTCMD.  There's an extra "ALLOW" parameter that causes the compile to fail.

![image](https://github.com/codefori/vscode-ibmi/assets/150696892/7d5d87b4-4da5-4e59-89c7-984cbbb60139)

CRTCMD CMD(MSIDRO/RTVIPLA) PGM(MSIDRO/RTVIPLA) SRCFILE(MSIDRO/QCMDSRC) ALLOW(ALLOW(*BPGM *IPGM *BMOD *IMOD) ) CURLIB(*NOCHG) PRDLIB(*NOCHG)

Thanks!

**Reply by worksofliam** [reply]
@b-smith-23 Where did you find that?

**Reply by b-smith-23** [reply]
If you do a cntrl + E from the source member you are given a list of create commands.  This one is specifically for allowing return variables.
